File: app/rag/indexer.py
from __future__ import annotations

import logging
import time
import uuid
from typing import Callable, Optional

from qdrant_client.models import FieldCondition, Filter, MatchValue, PointIdsList, PointStruct

from app.models import SourceDocument, PayloadKeys
from app.rag.chunker import TextChunker
from app.rag.embedding import EmbeddingService
from app.rag.vector_store import VectorStore

log = logging.getLogger(__name__)

_NAMESPACE = uuid.UUID("6ba7b810-9dad-11d1-80b4-00c04fd430c8")


def _make_id(doc_id: str) -> str:
    return str(uuid.uuid5(_NAMESPACE, doc_id))


def _make_chunk_id(doc_id: str, chunk_index: int) -> str:
    return str(uuid.uuid5(_NAMESPACE, f"{doc_id}_chunk_{chunk_index}"))


class DocumentIndexer:
    def __init__(
        self,
        vector_store: VectorStore,
        embedding: EmbeddingService,
        chunker: TextChunker,
    ) -> None:
        self._store = vector_store
        self._embedding = embedding
        self._chunker = chunker

    async def index(self, collection: str, doc: SourceDocument) -> None:
        await self.index_batch(collection, [doc])

    async def index_batch(
        self,
        collection: str,
        docs: list[SourceDocument],
        progress_cb: Optional[Callable[[int, int], None]] = None,
        replace_source_id: Optional[str] = None,
    ) -> None:
        now_ms = int(time.time() * 1000)

        # ── Phase 1: embed everything ──────────────────────────────────────────
        # Collection creation is deferred until after the first embed so we can
        # use the model's *actual* output dimension rather than the configured
        # one (e.g. Ollama models have a fixed output size the API cannot override).
        # No Qdrant writes happen here. If any embed call fails the exception
        # propagates immediately and nothing has been written, so no cleanup
        # is needed.
        points: list[PointStruct] = []

        for i, doc in enumerate(docs):
            chunks = self._chunker.chunk(doc.text)
            total_chunks = len(chunks)

            for chunk in chunks:
                vector = await self._embedding.embed(chunk.text)
                point_id = _make_id(doc.id) if total_chunks == 1 else _make_chunk_id(doc.id, chunk.index)

                payload: dict = {
                    PayloadKeys.TEXT: chunk.text,
                    PayloadKeys.INDEXED_AT_MS: now_ms,
                    PayloadKeys.SOURCE_DOC_ID: doc.id,
                    PayloadKeys.CHUNK_INDEX: str(chunk.index),
                    PayloadKeys.TOTAL_CHUNKS: str(total_chunks),
                }
                for k, v in doc.tags.items():
                    payload[PayloadKeys.tag(k)] = v
                for k, v in doc.properties.items():
                    payload[PayloadKeys.prop(k)] = v

                points.append(PointStruct(id=point_id, vector=vector, payload=payload))

            if progress_cb:
                progress_cb(i + 1, len(docs))

        if not points:
            return

        # Create the collection now that we know the actual vector size.
        if not await self._store.collection_exists(collection):
            actual_dims = len(points[0].vector)
            await self._store.create_collection(collection, dimensions=actual_dims)

        # ── Between phases: replace old vectors for this source ───────────────
        # Deleting here (after all embeds succeed, before any writes) means:
        # - embed failure → old vectors intact, nothing lost
        # - write failure → rollback removes the partial new batch
        if replace_source_id:
            filt = Filter(must=[
                FieldCondition(
                    key=PayloadKeys.tag("source_id"),
                    match=MatchValue(value=replace_source_id),
                )
            ])
            await self._store.delete_by_filter(collection, filt)

        # ── Phase 2: write to Qdrant with rollback on partial failure ──────────
        # Track every ID that was successfully committed so we can delete them
        # if a later batch throws.
        written_ids: list[str] = []
        try:
            batch_size = 100
            for i in range(0, len(points), batch_size):
                batch = points[i : i + batch_size]
                await self._store.upsert(collection, batch)
                written_ids.extend(str(p.id) for p in batch)
        except Exception:
            if written_ids:
                log.warning(
                    "Upsert failed after writing %d points — rolling back to Qdrant",
                    len(written_ids),
                )
                try:
                    await self._store.client.delete(
                        collection_name=collection,
                        points_selector=PointIdsList(points=written_ids),
                    )
                except Exception:
                    log.exception("Rollback delete also failed — collection %s may have orphaned points", collection)
            raise

File: internal/rag/indexer.go
// Document indexer — Go port of app/rag/indexer.py, including the 3-phase
// embed → replace → write-with-rollback protocol and deterministic UUIDv5
// point IDs (same namespace as Python, so re-indexing the same doc from
// either runtime produces the same point ID).
package rag

import (
	"context"
	"log"
	"strconv"
	"sync"
	"time"

	"github.com/google/uuid"

	"github.com/MichalOndrejka/conduit/internal/models"
)

// idNamespace matches _NAMESPACE in app/rag/indexer.py (uuid.NAMESPACE_DNS).
var idNamespace = uuid.MustParse("6ba7b810-9dad-11d1-80b4-00c04fd430c8")

func makeID(docID string) string {
	return uuid.NewSHA1(idNamespace, []byte(docID)).String()
}

func makeChunkID(docID string, chunkIndex int) string {
	return uuid.NewSHA1(idNamespace, []byte(docID+"_chunk_"+strconv.Itoa(chunkIndex))).String()
}

type DocumentIndexer struct {
	store     *VectorStore
	embedding *EmbeddingService
	chunker   *TextChunker

	// collMu serializes the "ensure collection matches this dimension" check
	// below. Multiple sources of the same type share a collection and can
	// sync concurrently (SyncSelected) — without this, two syncs can both
	// see a dimension mismatch and race on delete+recreate, one losing its
	// CreateCollection to a 409 "already exists".
	collMu sync.Mutex
}

func NewDocumentIndexer(store *VectorStore, embedding *EmbeddingService, chunker *TextChunker) *DocumentIndexer {
	return &DocumentIndexer{store: store, embedding: embedding, chunker: chunker}
}

// IndexBatchOptions carries the optional knobs of index_batch.
type IndexBatchOptions struct {
	// ProgressCb is called after each document is embedded: (done, total).
	ProgressCb func(done, total int)
	// ReplaceSourceID deletes existing vectors tagged with this source_id
	// after all embeds succeed and before any writes.
	ReplaceSourceID string
	// Checkpoint is called before each document; returning an error (e.g.
	// syncctl.ErrSyncCancelled) aborts the run before any Qdrant write.
	Checkpoint func() error
}

func (d *DocumentIndexer) Index(ctx context.Context, collection string, doc models.SourceDocument) error {
	return d.IndexBatch(ctx, collection, []models.SourceDocument{doc}, IndexBatchOptions{})
}

func (d *DocumentIndexer) IndexBatch(
	ctx context.Context, collection string, docs []models.SourceDocument, opts IndexBatchOptions,
) error {
	nowMs := time.Now().UnixMilli()

	// ── Phase 1: embed everything ───────────────────────────────────────────
	// Collection creation is deferred until after the first embed so we can
	// use the model's *actual* output dimension rather than the configured
	// one. No Qdrant writes happen here — an embed failure leaves everything
	// untouched.
	var points []Point

	for i, doc := range docs {
		if opts.Checkpoint != nil {
			if err := opts.Checkpoint(); err != nil {
				return err
			}
		}
		chunks := d.chunker.Chunk(doc.Text)
		totalChunks := len(chunks)

		for _, chunk := range chunks {
			vector, err := d.embedding.Embed(ctx, chunk.Text)
			if err != nil {
				return err
			}
			pointID := makeChunkID(doc.ID, chunk.Index)
			if totalChunks == 1 {
				pointID = makeID(doc.ID)
			}

			payload := map[string]any{
				models.PayloadText:        chunk.Text,
				models.PayloadIndexedAtMs: nowMs,
				models.PayloadSourceDocID: doc.ID,
				models.PayloadChunkIndex:  strconv.Itoa(chunk.Index),
				models.PayloadTotalChunks: strconv.Itoa(totalChunks),
			}
			for k, v := range doc.Tags {
				payload[models.TagKey(k)] = v
			}
			for k, v := range doc.Properties {
				payload[models.PropKey(k)] = v
			}

			points = append(points, Point{ID: pointID, Vector: vector, Payload: payload})
		}

		if opts.ProgressCb != nil {
			opts.ProgressCb(i+1, len(docs))
		}
	}

	if len(points) == 0 {
		return nil
	}

	// Create the collection now that we know the actual vector size, or
	// recreate it if an existing collection was sized for a different
	// embedding model — its vectors are in the wrong space and unusable
	// regardless of source, so every source touching it must be re-synced.
	size := len(points[0].Vector)
	if err := func() error {
		d.collMu.Lock()
		defer d.collMu.Unlock()
		if !d.store.CollectionExists(ctx, collection) {
			return d.store.CreateCollection(ctx, collection, size)
		}
		existing, err := d.store.VectorSize(ctx, collection)
		if err != nil {
			return err
		}
		if existing == size {
			return nil
		}
		log.Printf("collection %s vector size %d != %d — recreating for new embedding model", collection, existing, size)
		if err := d.store.DeleteCollection(ctx, collection); err != nil {
			return err
		}
		return d.store.CreateCollection(ctx, collection, size)
	}(); err != nil {
		return err
	}

	// ── Between phases: replace old vectors for this source ────────────────
	// Deleting here (after all embeds succeed, before any writes) means:
	//   - embed failure → old vectors intact, nothing lost
	//   - write failure → rollback removes the partial new batch
	if opts.ReplaceSourceID != "" {
		filter := &Filter{Must: []FieldCondition{{
			Key: models.TagKey("source_id"), Match: Match{Value: opts.ReplaceSourceID},
		}}}
		if err := d.store.DeleteByFilter(ctx, collection, filter); err != nil {
			return err
		}
	}

	// ── Phase 2: write to Qdrant with rollback on partial failure ──────────
	var writtenIDs []string
	const batchSize = 100
	for i := 0; i < len(points); i += batchSize {
		end := min(i+batchSize, len(points))
		batch := points[i:end]
		if err := d.store.Upsert(ctx, collection, batch); err != nil {
			if len(writtenIDs) > 0 {
				log.Printf("warning: upsert failed after writing %d points — rolling back", len(writtenIDs))
				if rbErr := d.store.DeleteByIDs(ctx, collection, writtenIDs); rbErr != nil {
					log.Printf("error: rollback delete also failed — collection %s may have orphaned points: %v", collection, rbErr)
				}
			}
			return err
		}
		for _, p := range batch {
			writtenIDs = append(writtenIDs, p.ID)
		}
	}
	return nil
}
